File: tg.js
const express = require('express');
const axios = require('axios');
const { HttpsProxyAgent } = require('https-proxy-agent');

const app = express();
const PORT = 3000;

// Настройка вашего прокси
const proxyUrl = 'http://api6427e610fa202b13_c_US_s_1:aepKxOZdTRMDH3XC@gate.node-proxy.com:10000';
const agent = new HttpsProxyAgent(proxyUrl);

// Middleware для парсинга JSON и URL-encoded данных
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// Логирование всех запросов
app.use((req, res, next) => {
    console.log(`[${new Date().toISOString()}] ${req.method} ${req.url}`);
    next();
});

// Корневой путь - информация о сервисе
app.get('/', (req, res) => {
    res.json({
        status: 'ok',
        message: 'Telegram proxy is running',
        proxy: proxyUrl,
        endpoints: {
            'GET /bot/:token/:method': 'Get bot info',
            'POST /bot/:token/:method': 'Send message',
            'GET /bot/:token/getMe': 'Get bot info'
        }
    });
});

// Универсальный обработчик для всех методов
app.all('/bot/:token/:method', async (req, res) => {
    try {
        const { token, method } = req.params;

        // Проверка наличия токена и метода
        if (!token || !method) {
            return res.status(400).json({
                error: 'Missing token or method',
                message: 'Expected /bot/<TOKEN>/<METHOD>'
            });
        }

        const url = `https://api.telegram.org/bot${token}/${method}`;

        console.log(`🔗 Проксируем запрос к: ${url}`);
        console.log(`📦 Метод: ${req.method}`);
        console.log(`📦 Параметры:`, req.query);
        console.log(`📦 Тело:`, req.body);

        // Настройка запроса к Telegram
        const config = {
            method: req.method,
            url: url,
            httpsAgent: agent,
            headers: {
                'Content-Type': 'application/json'
            },
            timeout: 30000,
            params: req.query // Для GET запросов
        };

        // Для POST запросов добавляем тело
        if (req.method === 'POST' || req.method === 'PUT' || req.method === 'PATCH') {
            config.data = req.body;
        }

        const response = await axios(config);

        console.log(`✅ Ответ получен, статус: ${response.status}`);
        res.json(response.data);

    } catch (error) {
        console.error(`❌ Ошибка:`, error.message);

        if (error.response) {
            // Ошибка от Telegram API
            console.error(`   Статус: ${error.response.status}`);
            console.error(`   Данные:`, error.response.data);
            res.status(error.response.status).json(error.response.data);
        } else if (error.request) {
            // Нет ответа от Telegram
            console.error(`   Нет ответа от Telegram API`);
            res.status(502).json({
                error: 'No response from Telegram API',
                message: error.message
            });
        } else {
            // Ошибка настройки запроса
            console.error(`   Ошибка настройки запроса:`, error.message);
            res.status(500).json({
                error: 'Internal proxy error',
                message: error.message
            });
        }
    }
});

// Обработка 404
app.use((req, res) => {
    res.status(404).json({
        error: 'Not found',
        message: `Endpoint ${req.url} not found`
    });
});

// Обработка ошибок
app.use((err, req, res, next) => {
    console.error('❌ Необработанная ошибка:', err);
    res.status(500).json({
        error: 'Internal server error',
        message: err.message
    });
});

app.listen(PORT, '0.0.0.0', () => {
    console.log(`\n🚀 Telegram proxy запущен на порту ${PORT}`);
    console.log(`🌐 Прокси сервер: ${proxyUrl}`);
    console.log(`\n📝 Тестовые команды:`);
    console.log(`  curl "http://localhost:${PORT}/bot/8878543727:AAEMchGTmSMZ95EpdINcg52G_J08N-uV58Q/getMe"`);
    console.log(`  curl -X POST "http://localhost:${PORT}/bot/8878543727:AAEMchGTmSMZ95EpdINcg52G_J08N-uV58Q/sendMessage" -H "Content-Type: application/json" -d '{"chat_id":"CHAT_ID","text":"Test"}'`);
    console.log(`\n📋 Health check: http://localhost:${PORT}/`);
});

File: tg.py
from fastapi import FastAPI, Request, Response
from fastapi.responses import JSONResponse
import httpx
import os
import re
import logging
from datetime import datetime
from dotenv import load_dotenv
import uvicorn

load_dotenv()

app = FastAPI()

# ============================================================
# Конфигурация
# ============================================================
PORT = int(os.getenv("PORT", 3000))
N8N_TARGET = os.getenv("N8N_TARGET", "https://n8n.w1do.ru").rstrip('/')
N8N_HOST = N8N_TARGET.split('://')[1].split('/')[0]

# Настройки прокси для исходящих запросов к Telegram
PROXY_HOST = os.getenv("PROXY_HOST")
PROXY_PORT = os.getenv("PROXY_PORT")
PROXY_USER = os.getenv("PROXY_USER")
PROXY_PASS = os.getenv("PROXY_PASS")

# Сборка прокси-URL
proxy_url = None
if PROXY_HOST and PROXY_PORT:
    if PROXY_USER and PROXY_PASS:
        proxy_url = f"http://{PROXY_USER}:{PROXY_PASS}@{PROXY_HOST}:{PROXY_PORT}"
    else:
        proxy_url = f"http://{PROXY_HOST}:{PROXY_PORT}"

# ============================================================
# Настройка логирования
# ============================================================
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# ============================================================
# Клиент для Telegram (с поддержкой прокси)
# ============================================================
telegram_client = httpx.AsyncClient(
    proxy=proxy_url,
    timeout=30.0,
    follow_redirects=False
)

# Клиент для n8n (без прокси, только внутренний)
n8n_client = httpx.AsyncClient(
    timeout=30.0,
    follow_redirects=False
)

# ============================================================
# Hop-by-hop заголовки (не проксируются)
# ============================================================
HOP_BY_HOP = {
    'host', 'connection', 'keep-alive', 'proxy-authenticate',
    'proxy-authorization', 'te', 'trailer', 'transfer-encoding',
    'upgrade', 'content-length'
}

# ============================================================
# Хелпер для очистки заголовков
# ============================================================
def clean_headers(headers: dict) -> dict:
    """Удаляет hop-by-hop заголовки и возвращает чистый dict"""
    cleaned = {}
    for key, value in headers.items():
        if key.lower() not in HOP_BY_HOP:
            cleaned[key] = value
    return cleaned

# ============================================================
# Мидлвари: логирование запросов
# ============================================================
@app.middleware("http")
async def log_requests(request: Request, call_next):
    start_time = datetime.now()
    response = await call_next(request)
    duration = (datetime.now() - start_time).total_seconds() * 1000
    logger.info(f"[{datetime.now().isoformat()}] {request.method} {request.url.path} - {response.status_code} ({duration:.0f}ms)")
    return response

# ============================================================
# Health check
# ============================================================
@app.get("/")
async def health_check():
    return {
        "status": "ok",
        "message": "Telegram proxy is running",
        "proxy": re.sub(r':[^:@]*@', ':****@', proxy_url) if proxy_url else "direct",
        "n8n": N8N_TARGET
    }

# ============================================================
# Проксирование запросов к Telegram Bot API
#   /bot/<TOKEN>/<METHOD>  ->  https://api.telegram.org/bot<TOKEN>/<METHOD>
# ============================================================
@app.api_route("/bot/{token:path}/{method:path}", methods=["GET", "POST", "PUT", "PATCH", "DELETE"])
async def proxy_to_telegram(token: str, method: str, request: Request):
    try:
        url = f"https://api.telegram.org/bot{token}/{method}"

        # Получаем тело запроса
        body = await request.body()

        # Проксируем ВСЕ заголовки от n8n к Telegram
        headers = clean_headers(dict(request.headers))
        headers['host'] = 'api.telegram.org'  # Важно для правильной маршрутизации

        # ---------- ДИАГНОСТИКА ----------
        logger.info("========== -> TELEGRAM REQUEST ==========")
        logger.info(f"URL: {url}")
        logger.info(f"Method: {request.method}")
        logger.info(f"Headers: {headers}")
        if body:
            try:
                logger.info(f"Body: {body.decode('utf-8')[:500]}")
            except Exception:
                logger.info(f"Body: (binary, {len(body)} bytes)")
        logger.info("========================================")

        # Отправляем запрос в Telegram через прокси
        response = await telegram_client.request(
            method=request.method,
            url=url,
            headers=headers,
            content=body,  # Передаем сырое тело
            params=dict(request.query_params)
        )

        # ---------- ДИАГНОСТИКА ----------
        logger.info("========== <- TELEGRAM RESPONSE ==========")
        logger.info(f"Status: {response.status_code}")
        logger.info(f"Headers: {response.headers}")
        if response.content:
            try:
                preview = response.text[:500]
                logger.info(f"Body: {preview}")
            except Exception:
                logger.info(f"Body: (binary, {len(response.content)} bytes)")
        logger.info("==========================================")

        # Проксируем ответ от Telegram обратно в n8n
        resp_headers = clean_headers(dict(response.headers))

        return Response(
            content=response.content,
            status_code=response.status_code,
            headers=resp_headers
        )

    except Exception as e:
        logger.error(f"❌ Telegram error: {str(e)}")
        return JSONResponse(
            status_code=500,
            content={"error": str(e)}
        )

# ============================================================
# Проксирование вебхуков в n8n
#   POST /webhook-test/<uuid>/webhook -> POST {N8N}/webhook-test/<uuid>/webhook
#   POST /webhook/<uuid>/webhook      -> POST {N8N}/webhook/<uuid>/webhook
# ============================================================
@app.api_route("/webhook-test/{path:path}", methods=["GET", "POST", "PUT", "PATCH", "DELETE"])
@app.api_route("/webhook/{path:path}", methods=["GET", "POST", "PUT", "PATCH", "DELETE"])
async def forward_to_n8n(request: Request, path: str):
    target_url = f"{N8N_TARGET}{request.url.path}"

    # Получаем сырое тело (сохраняем как есть, без парсинга)
    body = await request.body()

    # Проксируем ВСЕ заголовки от Telegram в n8n
    headers = clean_headers(dict(request.headers))
    headers['host'] = N8N_HOST

    # ---------- ДИАГНОСТИКА ----------
    logger.info("========== -> N8N REQUEST ==========")
    logger.info(f"URL: {target_url}")
    logger.info(f"Method: {request.method}")
    logger.info(f"Headers: {headers}")
    if body:
        try:
            logger.info(f"Body: {body.decode('utf-8')[:500]}")
        except Exception:
            logger.info(f"Body: (binary, {len(body)} bytes)")
    logger.info("====================================")

    try:
        # Отправляем в n8n
        response = await n8n_client.request(
            method=request.method,
            url=target_url,
            headers=headers,
            content=body,
            params=dict(request.query_params)  # query параметры тоже пробрасываем
        )

        # ---------- ДИАГНОСТИКА ----------
        logger.info("========== <- N8N RESPONSE ==========")
        logger.info(f"Status: {response.status_code}")
        logger.info(f"Headers: {response.headers}")
        if response.content:
            try:
                preview = response.text[:500]
                logger.info(f"Body: {preview}")
            except Exception:
                logger.info(f"Body: (binary, {len(response.content)} bytes)")
        logger.info("=====================================")

        # Проксируем ответ от n8n в Telegram
        resp_headers = clean_headers(dict(response.headers))

        return Response(
            content=response.content,
            status_code=response.status_code,
            headers=resp_headers
        )

    except Exception as e:
        logger.error(f"[N8N PROXY ERROR] {str(e)}")
        return JSONResponse(
            status_code=502,
            content={"error": "Bad gateway to n8n", "detail": str(e)}
        )

# ============================================================
# 404 для всех остальных путей
# ============================================================
@app.api_route("/{path:path}", methods=["GET", "POST", "PUT", "PATCH", "DELETE"])
async def not_found(request: Request, path: str):
    return JSONResponse(status_code=404, content={"error": "Not found"})

# ============================================================
# Запуск
# ============================================================
if __name__ == "__main__":
    print(f"🚀 Proxy server started on port {PORT}")
    print(f"🌐 Outbound proxy (Telegram): {'****' if proxy_url else 'direct'}")
    print(f"🎯 n8n target: {N8N_TARGET}")

    uvicorn.run(
        app,
        host="0.0.0.0",
        port=PORT,
        log_level="info"
    )
